src/libAtomVM/utils.h

```diff
@@ -102,23 +102,53 @@
         #define ENDIAN_SWAP_32(value) ((((value) & 0xFF) << 24) | (((value) & 0xFF00) << 8) | (((value) & 0xFF0000) >> 8) | (((value) & 0xFF000000) >> 24))
     #endif
 
-#else
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+
+    #define READ_64_UNALIGNED(ptr) \
+        ( (((uint64_t) ((uint8_t *)(ptr))[0]) << 56) | (((uint64_t) ((uint8_t *) (ptr))[1]) << 48) | \
+            (((uint64_t) ((uint8_t *)(ptr))[2]) << 40) | (((uint64_t) ((uint8_t *) (ptr))[3]) << 32) | \
+            (((uint64_t) ((uint8_t *)(ptr))[4]) << 24) | (((uint64_t) ((uint8_t *) (ptr))[5]) << 16) | \
+            (((uint64_t) ((uint8_t *)(ptr))[6]) << 8) | (((uint64_t) ((uint8_t *) (ptr))[7])) )
+
+    #define WRITE_64_UNALIGNED(ptr, val) \
+        { \
+            ((uint8_t *)(ptr))[0] = (((uint64_t) val) >> 56) & 0xff; \
+            ((uint8_t *)(ptr))[1] = (((uint64_t) val) >> 48) & 0xff; \
+            ((uint8_t *)(ptr))[2] = (((uint64_t) val) >> 40) & 0xff; \
+            ((uint8_t *)(ptr))[3] = (((uint64_t) val) >> 32) & 0xff; \
+            ((uint8_t *)(ptr))[4] = (((uint64_t) val) >> 24) & 0xff; \
+            ((uint8_t *)(ptr))[5] = (((uint64_t) val) >> 16) & 0xff; \
+            ((uint8_t *)(ptr))[6] = (((uint64_t) val) >> 8) & 0xff; \
+            ((uint8_t *)(ptr))[7] = ((uint64_t) val) & 0xff; \
+        }
+
```

Comment:
IIRC, these macros are designed to read and write (64, 32, 16 bit)  integer values in network byte order (big-endian, so the high-order byte is written first in s sequence on media — whether it’s disk, socket, etc).

The big and little endian versions of these macros appear to be the same in this PR.  Does this work because we are using the array indexing feature of C to place the bytes in the right location?  (Sorry, my C details are rusty)

Reply:
They are all the same because they work in the same way, what really matters is the write target endianess.
This code is quite messy and there are duplicated and slow things, and I'm alrady working to replace it.
Also the macros here are not easy to understand, and there is also duplication with bitstring.h code.

So my plan was to just fix this, and to focus on a wider and cleaner rework later.

Reply:
Yeah I was just curious about the C semantics.